### PR TITLE
Improve approx scalar repr method for better readability 

### LINKED ISSUE #6985: pytest.approx: Please improve repr readability
It seems that pytest.approx repr is optimized for very high or very low tolerances, but is not as easily readable as it could be for values in between.
Some examples:

- `pytest.approx(42, abs=1)` repr is `42 ± 1.0e+00`, which could be just `42 ± 1`.
- `pytest.approx(5, rel=0.01)` repr is `5 ± 5.0e-02`, but could be `5 ± 0.05` or maybe even `5 ± 1%`
- `pytest.approx(24000, abs=500)` repr is `24000 ± 5.0e+02` and could be `24000 ± 500`
- removing accuracy example: `pytest.approx(1500, abs=555)` repr is `1500 ± 5.6e+02` and could be `1500 ± 555`

In these number ranges using the exact number instead of the exponential notation is both more accurate **and** easier readable for humans. Please adapt pytest.approx to switch it's repr style according to the input value ranges.

### PR DESCRIPTION
**Pull request by fazeelghafoor**
Closes #6985 

- [ ] Include documentation when adding new features.
- [X] Include new tests or update existing tests when applicable.
- [X] Allow maintainers to push and squash when merging my commits. Please uncheck this if you prefer to squash the commits yourself.
- [X] Add text like ``closes #XYZW`` to the PR description and/or commits (where ``XYZW`` is the issue number).
- [X] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`.
- [ ] Add yourself to `AUTHORS` in alphabetical order.

### CODE REVIEW

**Comment by fazeelghafoor** [comment]
@Zac-HD please review and let me know if there are any changes required.

**Review by Pierre-Sassoulas — COMMENTED** [review]
Thank you for the pull request !

Did you see this comment in the original issue : https://github.com/pytest-dev/pytest/issues/6985#issuecomment-606326803 ?

**Comment by fazeelghafoor** [comment]
@Pierre-Sassoulas thank you for the feedback. I had missed the comment so I have added the 'n' format specifier as well. Please let me know what else needs to changed

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Some tests results to upgrade otherwise LGTM

**Comment by fazeelghafoor** [comment]
added test cases as well

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
The pipeline is still failing, some existing tests need to be upgraded too.

**Comment by fazeelghafoor** [comment]
@Pierre-Sassoulas updated existing tests

File: src/_pytest/runner.py [hunk 1]
```diff
@@ -177,7 +177,7 @@ def pytest_runtest_call(item: Item) -> None:
         sys.last_type = type(e)
         sys.last_value = e
         if sys.version_info >= (3, 12, 0):
-            sys.last_exc = e
+            sys.last_exc = e  # type: ignore[attr-defined]
```
Comment (Pierre-Sassoulas) [inline src/_pytest/runner.py:180]: Seems like this type ignores are unused in the CI (see pre-commit job).

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Great, do you mind adding a changelog too, please ?

File: changelog/6985.improvement.rst [hunk 2]
```diff
@@ -0,0 +1,15 @@
+Improved the `ApproxScalar` class to enhance the readability of the `repr` method for intermediate ranges of values and tolerances.
```
Comment (nicoddemus) [inline changelog/6985.improvement.rst:1]: While it is technically correct, `ApproxScalar` is internal so we probably should phrase this with `pytest.approx`, which is what users actually use:

```suggestion
Improved :func:`pytest.approx` to enhance the readability of intermediate value ranges and tolerances.
```

File: changelog/6985.improvement.rst [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  Example:
+  **Previous Output:**
+  .. code-block:: console
+      >>> pytest.approx(42, abs=1)
+      42 ± 1.0e+00
+
+  **Current Output:**
+  .. code-block:: console
+      >>> pytest.approx(42, abs=1)
+      42 ± 1
```
Comment (nicoddemus) [inline changelog/6985.improvement.rst:13]: The previous example was correctly formatted, this should fix it:

```suggestion

    Example:
  
    **Previous Output:**
    
    .. code-block:: console
    
        >>> pytest.approx(42, abs=1)
        42 ± 1.0e+00

    **Current Output:**
    
    .. code-block:: console
    
        >>> pytest.approx(42, abs=1)
        42 ± 1
```

File: src/_pytest/python_api.py [hunk 4]
```diff
@@ -406,7 +406,11 @@ def __repr__(self) -> str:
         # If a sensible tolerance can't be calculated, self.tolerance will
         # raise a ValueError.  In this case, display '???'.
         try:
-            vetted_tolerance = f"{self.tolerance:.1e}"
+            if self.tolerance >= 1e-3 and self.tolerance < 1e3:
```
Comment (nicoddemus) [inline src/_pytest/python_api.py:409]: Same thing, but a bit more Pythonic:

```suggestion
            if 1e-3 <= self.tolerance < 1e3:
```

**Review by nicoddemus — APPROVED** [review]
LGTM, left some minor comments.

**Comment by fazeelghafoor** [comment]
@nicoddemus thanks for the suggestions. can you let me know what could be causing the readthedocs check to fail. I don't understand the error

**Comment by The-Compiler** [comment]
Looks like an issue unrelated to your PR: https://github.com/sphinx-contrib/sphinxcontrib-towncrier/issues/92

**Comment by nicoddemus** [comment]
> can you let me know what could be causing the readthedocs check to fail. I don't understand the error

It might be some regression on towncrier/sphinx extension, I noticed there was a new release towncrier release 5 hrs ago:

https://pypi.org/project/towncrier/#history

Don't worry, seems unrelated to your changes.

**Review by The-Compiler — APPROVED** [review]
Looking great!

**Comment by nicoddemus** [comment]
Opened https://github.com/pytest-dev/pytest/pull/12675 to fix the readthedocs/towncrier issue, we can update with `main` once that lands.

File: changelog/6985.improvement.rst [hunk 5]
```diff
@@ -0,0 +1,21 @@
+Improved :func:`pytest.approx` to enhance the readability of intermediate value ranges and tolerances.
```
Comment (Pierre-Sassoulas) [inline changelog/6985.improvement.rst:1]:
```suggestion
Improved :func:`pytest.approx` to enhance the readability of values ranges and tolerances between 0.001 and 1000.
```

Maybe it's clearer to be explicit and remove 'intermediate' ? Or define it at the beggining:


```suggestion
Improved :func:`pytest.approx` to enhance the readability of intermediate value ranges and tolerances (between 0.001 and 1000).
```

Comment (nicoddemus) [reply changelog/6985.improvement.rst:1]: I like the first one, more concise, but either one is OK with me.

File: changelog/6985.improvement.rst [hunk 6]
```diff
@@ -0,0 +1,21 @@
+Improved :func:`pytest.approx` to enhance the readability of value ranges and tolerances between 0.001 and 1000.
+  * The `repr` method now provides clearer output for values within intermediate ranges, making it easier to interpret the results.
```
Comment (Pierre-Sassoulas) [inline changelog/6985.improvement.rst:2]:
```suggestion
  * The `repr` method now provides clearer output for values within those ranges, making it easier to interpret the results.
```

File: changelog/6985.improvement.rst [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  * Previously, the output for intermediate ranges of values and tolerances was displayed in scientific notation (e.g., `42 ± 1.0e+00`). The updated method now presents the tolerance as a decimal for better readability (e.g., `42 ± 1`).
```
Comment (Pierre-Sassoulas) [inline changelog/6985.improvement.rst:3]:
```suggestion
  * Previously, the output for those ranges of values and tolerances was displayed in scientific notation (e.g., `42 ± 1.0e+00`). The updated method now presents the tolerance as a decimal for better readability (e.g., `42 ± 1`).
```

**Review by Pierre-Sassoulas — APPROVED** [review]

**Comment by Pierre-Sassoulas** [comment]
Great !

**Review by nicoddemus — APPROVED** [review]

**Pull request merged**